Source/Scene/VoxelPrimitive.js

```diff
@@ -0,0 +1,2514 @@
+ */
+function getGlslField(type, index) {
+  if (type === MetadataType.SCALAR) {
+    return "";
+  }
+  return `.[${index}]`;
+}
+
+/**
+ * @param {VoxelPrimitive} that
+ * @param {Context} context
+ * @private
+ */
+function buildDrawCommands(that, context) {
+  const provider = that._provider;
+  const shapeType = provider.shape;
+  const names = provider.names;
+  const types = provider.types;
+  const componentTypes = provider.componentTypes;
+  const depthTest = that._depthTest;
+  const useLogDepth = that._useLogDepth;
+  const paddingBefore = that.paddingBefore;
+  const paddingAfter = that.paddingAfter;
+  const minBounds = that._minBounds;
+  const maxBounds = that._maxBounds;
+  const minimumValues = provider.minimumValues;
+  const maximumValues = provider.maximumValues;
+  const minClippingBounds = that._minClippingBounds;
+  const maxClippingBounds = that._maxClippingBounds;
+  const keyframeCount = that._keyframeCount;
+  const despeckle = that._despeckle;
+  const jitter = that._jitter;
+  const nearestSampling = that._nearestSampling;
+  const customShader = that._customShader;
+  const attributeLength = types.length;
+
+  const shaderBuilder = new ShaderBuilder();
+
+  // Vertex shader
+  shaderBuilder.addVertexLines(["#line 0", VoxelVS]);
+  shaderBuilder.setPositionAttribute("vec4", "a_position");
+
+  // Fragment shader
+  shaderBuilder.addFragmentLines([
+    customShader.fragmentShaderText,
+    "#line 0",
+    VoxelFS,
+  ]);
+
+  const fragmentStructId = "FragmentInput";
+  const fragmentStructName = "FragmentInput";
+  const attributeStructId = "Attributes";
+  const attributeStructName = "Attributes";
+  const attributeFieldName = "attributes";
+  const voxelStructId = "Voxel";
+  const voxelStructName = "Voxel";
+  const voxelFieldName = "voxel";
+
+  // Attribute struct
+  shaderBuilder.addStruct(
+    attributeStructId,
+    attributeStructName,
+    ShaderDestination.FRAGMENT
+  );
+
+  for (let i = 0; i < attributeLength; i++) {
+    const name = names[i];
+    const type = types[i];
+    const glslType = getGlslType(type);
+    shaderBuilder.addStructField(attributeStructId, glslType, name);
+  }
+
+  // Voxel struct
+  shaderBuilder.addStruct(
+    voxelStructId,
+    voxelStructName,
+    ShaderDestination.FRAGMENT
+  );
+
+  for (let i = 0; i < attributeLength; i++) {
+    const name = names[i];
+    const type = types[i];
+    const glslType = getGlslType(type);
+    shaderBuilder.addStructField(voxelStructId, glslType, `${name}Minimum`);
+    shaderBuilder.addStructField(voxelStructId, glslType, `${name}Maximum`);
+    shaderBuilder.addStructField(voxelStructId, "vec3", `${name}NormalLocal`);
+    shaderBuilder.addStructField(voxelStructId, "vec3", `${name}NormalWorld`);
+    shaderBuilder.addStructField(voxelStructId, "vec3", `${name}NormalView`);
+    shaderBuilder.addStructField(voxelStructId, "bool", `${name}NormalValid`);
+  }
+
+  shaderBuilder.addStructField(voxelStructId, "vec3", "positionEC");
```

Comment:
the Model equivalent also allows `positionMC` and `positionWC` (albeit the latter is not super useful unless we can give it more precision somehow...). Though voxels could be different.

Definitely check the https://github.com/CesiumGS/cesium/tree/main/Documentation/CustomShaderGuide. Maybe voxels need their own section for differences.